### DISCUSSION TITLE
Add brace collapsing to E2 editor

### DISCUSSION THREAD

**Discussion by ghost** [wiremod/wire#2632] [Suggestions]
Functionality is same as hiding functions, etc., but realization must be simple, I think
1. "+" button after number for each line, containing '{' or #[
2. when pressed, remember current formatting (spaces, newlines) and crop them to just single spaces, making whole content one-line.
3. when clicked on that line, return formatting to edit and set button to '-'

That would definitely make a lot of people writing loooong e2's happy
- I know about #include, but, taking in consideration e2's limitations, it's only usable for small, maybe for average-size e2's. When u write 5000 lines and must always think about size limit (about 227KB), includes only hurts.
  ![init_end](https://cloud.githubusercontent.com/assets/10134635/5366543/bc97f96e-8005-11e4-8ca0-8cd23e0e00ed.png)

**Comment by Divran** [comment]
Includes have a size limit? I don't think there's a limit set anywhere in the code.

**Comment by ghost** [comment]
there may be no limit in e2, BUT there is limit on e2 file, which can be transferred to server. If you try to upload e2 sized over around 227KB, your client will FREEZE, spamming in console, something about file can't be transfered, I don;t remember exactly. That "freezing" means you can not do anything, only disconnect from server and connect again, it will freeze even in single player. And includes works just exactly like copy&paste, so it doesnt matter if you have 1 very large file or 10 smaller ones, anyways same large file will be generated and transfered. I am currently writing E2 PC (the most complex ever in gmod, I think) and that tons of limitations are really pissing me off. It takes about 20 seconds to just find line in one place, than same in another, argh...

**Comment by ghost** [comment]
@Divran I was thinking, it also may be very helpful to have comments like "#_+label something" and "#_!label", so that first can be link to position of second, supporting multiple links. Well, this is a bit weird, but useful. And links from functions to definitions...

**Comment by Divran** [comment]
So basically you have a file so large that you're hitting Lua's performance limit.

**Comment by ghost** [comment]
I never thought about it like that. It was more likely transfer limit, which can be 256KB - some meta deta included, but maybe. Do I need to test that behavior again to say correctly and add a new issue?

**Comment by Divran** [comment]
I don't think there is a transfer limit though, because the code splits up the transmitting over time automatically. The only explanation I have is that your code is simply too big for other reasons. For example, you said that searching through the file (with ctrl+f I assume?) takes several seconds. That's a hint.

Hmm.. but you also said your client freezes when sending. Are you sure you're using the latest version of wiremod? In an older version, the splitting up of transmissions of E2 files wasn't very well coded, and that might cause the freezing you are seeing.

**Comment by ghost** [comment]
No-no, I didn't mean ctrl-f search, I meant just looking around to find what I need, and auto-search is still pretty fast, less than second.

Yeah, It was actual about half a year ago, not for last version. 
I'll try today and check if that is still actual problem.

**Comment by Divran** [comment]
uhhh yeah testing with the latest version before you post a bug report is generally a good idea

**Comment by ghost** [comment]
tested a little, but hadn't enough time to determine the limit, and forgot, what it spammed to console again... lol.. but this problem is still actual, I think it's exactly same as it was. I'll check further later

**Comment by ghost** [comment]
So, it's gone to off-topic, but what status is it on topic?

**Comment by thegrb93** [comment]
ExpAdv2 will have this feature soon. Just ask the creator if you can use that code for this editor, or if you can replace this editor with theirs.

**Comment by Divran** [comment]
Can't simply copy paste code from a completely different editor. That's not how it works. (At least not for  such a duct taped editor like E2s)

**Comment by vurvdev** [comment]
StarfallEx has this feature on their fork of the wire editor. Could ask the PR creator if their code could be brought into the original wire editor.
